sd_webui_bayesian_merger/merger.py

```diff
@@ -160,6 +176,8 @@ def merge_block(
             return (1 - alpha - beta) * t0 + alpha * t1 + beta * t2
         elif self.cfg.merge_mode == "weighted_sum":
             return (1 - alpha) * t0 + alpha * t1
+        elif self.cfg.merge_mode == "double_diff":
+            return t0 + alpha * (t1 - t2 + t3 - t4)
```

Comment:
Hmmm, why not this?

```suggestion
            return t0 + alpha * (t1 - t2) + beta * (t3 - t4)
```

If the goal is really to add (t1 - t2) and (t3 - t4) in equal amounts in any given block then I think you could just do a (global) 50-50 merge of those first, and then weighted_sum merge that into t0, right?

Reply:
good point

that's a placeholder for now, I'm still thinking about the different options.

For example, given that the classi `add_difference` is `(1-alpha)*A + alpha*(B-C)`, we can have:
- the one you are proposing:`A + alpha * (B - C) + beta * (D - E)`
- this would be the same of merging first the two differences with classic merge an the `add_difference` with `A`: `A + alpha * ( beta * (B - C) + gamma * (D - E) )`
- this is the one I like better, `(1-alpha)*A + alpha*( beta*(B-C) + gamma*(D-E) )` where we introduce a correction for `A` and we could (in theory) avoid saturating weights after one too many merges
- a variation of the last one, requiring only two parameters: `(1-alpha)*A + alpha*( (1 - beta)*(B-C) + beta*(D-E) )`

I may end up implementing all of them and let users experiment :)

Reply:
> * this would be the same of merging first the two differences with classic merge an the `add_difference` with `A`: `A + alpha * ( beta * (B - C) + gamma * (D - E) )`

I'm not sure this is exactly correct. Tensors themselves are linear, but we optimize a system of tensors and non-linear activation functions. So an optimal classic merge would not necessarily represent the optimal mixture to `add_difference`.

The question or tradeoff, I think, is how "off" the true optimal distribution would be from the one we'd predict if it were a truly linear system - and whether it's worth the cost of introducing the additional optimization parameter to account for that.

I was thinking about a feature idea the other day and I think this relates. Suppose I want to search for optimal mixtures within a neighborhood of a certain initial mixture? If I had such a method, then I could test the effects of finer perturbations / larger space of mixtures to search / simultaneous optimization of multiple mixtures vs. doing simpler merges in sequence.

Reply:
Actually something I noticed with triple sum is that it currently allows for negative weights to be applied to t0. Which could flip signs in the tensor and actually SUBTRACT t0 from the sum of t1 and t2. The current construction of triple sum is quite strange since if you break it out you would really want the equation to be doing either of the following:
(1-beta)*((1-alpha)*t0+alpha*t1)+beta*t2 
Or
(1-alpha-beta+(alpha*beta))*t0+t1*alpha+beta*t2

Although this changes the behavior, this feel significantly more correct in terms of actual weighting, where as before you could end up with negative values in 1-alpha-beta. The first one just makes it so that the sum between t0 and t1 becomes weighted against beta, which some could argue actually makes alpha dependent on betas value. The second option keeps the independence of alpha and beta intact, but creates a weight that is derived from both values in such a way that it won't go negative. Some preliminary results with the second equation shows that the weight it produces is such that when alpha and beta are small, then the weight of t0 is small, and when alpha and beta are large then the weight of t0 is small, but if either alpha or beta reaches 1 then t0 is not applied, and only the sum between t1 and t2 is applied.

The second option was making the weight that is applied to t0 equal to the product of 1-beta and 1-alpha. That way there will never be a negative result and in theory should respect the independence of both alpha and beta as they are applied to t1 and t2 respectively although the issue with this is that when either alpha or beta are near 1, then the impact of t0 could reach 0.

Graph of the second weight function can be found here: https://www.desmos.com/calculator/lzxtsebytt

Reply:
@mariaWitch 's triple merge formula is equivalent to the supermerger triple merge formula, _if you follow the proper coordinate inequalities when using the supermerger formula_. It really helps to look at it this way.

I. Start by deriving @mariaWitch 's merge formula iteratively by doing back to back merges (merge coefficients optimized jointly, not in sequence, which is the whole point of having a triple merge to begin with):

1. Start with t0

2. Merge some t1 in to get `(1 - alpha)*t0 + alpha*t1`, and call this T1

3. Merge some t2 into T1 to get `(1 - beta)*T1 + beta*t2` and then substitute and rearrange terms,

--> `(1 - beta)*[(1 - alpha)*t0 + alpha*t1] + beta*t2` [This is exactly @mariaWitch 's triple formula]

--> `(1 - alpha - beta + alpha*beta)*t0 + (alpha - alpha*beta)*t1 + beta*t2` [This is exactly @mariaWitch 's alternate version of the triple merge formula]

II. To see how supermerger got their formula, you can change coordinates in your search space using a new coordinate instead of alpha, `gamma = alpha - alpha*beta`, and get

--> `(1 - gamma - beta)*t0 + gamma*t1 + beta*t2`

Where now instead of searching in the `alpha, beta` coordinate system, we search in the `gamma, beta` coordinate system instead.

BUT, what combinations of coordinates are allowed in the `alpha, beta` and the `gamma, beta` coordinate systems? We have to look at the relationship between alpha and gamma when beta has some fixed value (i.e. by using the fixed beta in the change of variables equation we introduced with gamma). For example, if `beta = 0` then `gamma = alpha`; OTOH if `beta = 1` then `gamma = 0`; and in the general case, `gamma <= 1 - beta`

_The allowed range of values for gamma depends on the value of beta when using the supermerger formula_, and @s1dlx I think I recall seeing some concern about this at some point, but I think this should clear it up?

TLDR -- if your original search ( @mariaWitch style) in the alpha, beta coordinate system is within `0 <= alpha <= 1` and `0 <= beta <= 1` then , then your equivalent search (supermerger style) in the gamma, beta coordinate system must be within `0 <= beta <= 1` and `0 <= gamma <= (1 - beta)`

If you try some forbidden combination where `gamma > 1 - beta` instead, this is exactly what would give you a negative merge coefficient for t0. Either coordinate system on the search space should work, as long as you search within the right range of values for that coordinate system. In the @mariaWitch system, the equations have a bit more going on, but the range of values is simple. In the supermerger coordinate system, the merge equations look simpler but you have to be careful about what range of values you search.

Reply:
Note that in each case, we're only using two search space coordinates. For the supermerger formula I'm calling those `gamma, beta` to make the distinction from @mariaWitch 's `alpha`, but in practice you would just call them `alpha, beta`.

Overall, I think @mariaWitch 's way of doing it is probably more intuitive / interpretable, because end users can think of it as merging some combination of t0 and t1, and then merging some combination of that resulting model and t2, but delaying the optimization of merge coefficients until the end; i.e. it's just back-to-back sequential merges, but with the coefficients optimized jointly. This way of thinking about it also makes it easier to interpret the geometry of the search space (i.e. just a simple box region).